snuba/clickhouse/translators/snuba/mapping.py

```diff
@@ -115,27 +115,63 @@ def __init__(self, translation_rules: TranslationMappers) -> None:
             lambdas=[DefaultLambdaMapper()],
         )
         self.__translation_rules = translation_rules.concat(default_rules)
+        self.__cache: MutableMapping[Expression, Expression] = {}
 
     def visit_literal(self, exp: Literal) -> Expression:
-        return apply_mappers(exp, self.__translation_rules.literals, self)
+        if exp in self.__cache:
```

Comment:
So I had something that looked like this:
```
    def __use_cache(
        self,
        exp: Expression,
        rules: Sequence[
            ExpressionMapper[TExpIn, TExpOut, SnubaClickhouseStrictTranslator]
        ],
    ) -> Expression:
        if exp in self.__cache:
            return self.__cache[exp]

        ret = apply_mappers(exp, rules, self)
        assert isinstance(ret, Expression)  # mypy
        self.__cache[exp] = ret
        return ret

    def visit_literal(self, exp: Literal) -> Expression:
        return self.__use_cache(exp, self.__translation_rules.literals)
```
However, I would see this error on the `ret = apply_mappers...` line: `Cannot infer type argument 1 of "apply_mappers"`.

Reply:
```
def __use_cache(
        self,
        exp: TExpIn,
        rules: Sequence[
            ExpressionMapper[TExpIn, TExpOut, SnubaClickhouseStrictTranslator]
        ],
    ) -> Expression:
```
+
Changing the definition of `TexpIn` into `TExpIn=TypeVar("TExpin", bound=Expression)`


Should be enough.

Reply:
No, I get the same errors as before even with this change.

Reply:
I think I'll merge this change just to get the fix out and then we can circle back on fixing this because having this much duplicated code is driving me crazy.